src/msgpack.lua

```diff
@@ -0,0 +1,194 @@
+        for i=from,to,1 do
+            xlat_table[i] = filler
+        end
+    end
+    return xlat_table
+end
+
+local mp_type_hints = init_type_hints_array({
+    {0x80, 0x8f, builtin.MP_MAP}, -- MP_MAP (fixed)
+    {0x90, 0x9f, builtin.MP_ARRAY}, -- MP_ARRAY (fixed)
+    {0xdc, 0xdd, builtin.MP_ARRAY}, -- MP_ARRAY (16,32)
+    {0xde, 0xdf, builtin.MP_MAP}, -- MP_MAP (16,32)
+})
+
+local function mp_typeof(c)
+    return mp_type_hints[c] or builtin.MP_NIL
+end
+
+local intbuff
+local strict_alignment = (jit.arch == 'arm' or
+                          jit.arch == 'ppc' or
+                          jit.arch == 'mips')
+
+if strict_alignment then
+   intbuff = ffi.new('union intbuff[1]')
+end
+
+local function bswap_u16(num)
+    return bit.rshift(bit.bswap(tonumber(num)), 16)
+end
+
+local function decode_u16_risc(data)
+    ffi.copy(intbuff, data[0], 2)
+    data[0] = data[0] + 2
+    return tonumber(bswap_u16(intbuff[0].u16))
+end
+
+local function decode_u32_risc(data)
+    ffi.copy(intbuff, data[0], 4)
+    data[0] = data[0] + 4
+    return tonumber(ffi.cast('uint32_t', bit.bswap(tonumber(intbuff[0].u32))))
+end
+
+local function decode_u16_cisc(data)
+    local num = bswap_u16(ffi.cast(uint16_ptr_t, data[0])[0])
+    data[0] = data[0] + 2
+    return tonumber(num)
+end
+
+local function decode_u32_cisc(data)
+    local num = ffi.cast('uint32_t', bit.bswap(tonumber(ffi.cast(uint32_ptr_t, data[0])[0])))
+    data[0] = data[0] + 4
+    return tonumber(num)
+end
+
+local decode_u16
+local decode_u32
+
+if strict_alignment then
+    decode_u16 = decode_u16_risc
+    decode_u32 = decode_u32_risc
+else
+    decode_u16 = decode_u16_cisc
+    decode_u32 = decode_u32_cisc
+end
+
+local function decode_array_header_slowly(c, data)
+    -- fixarray: 0x90 .. 0x9f
+    if bit.band(c, 0xf0) == 0x90 then
+        return bit.band(c, 0xf)
+    end
+
+    -- array 16
+    if c == 0xdc then
+        return decode_u16(data)
+    -- array 32
+    elseif c == 0xdd then
+        return decode_u32(data)
+    else
+        error('Unsupported MP_ARRAY')
+    end
+end
+
+local function decode_array_header_compat(buf, size)
+    if type(buf) == "string" then
+        bufp[0] = ffi.cast(char_ptr_t, buf)
+    elseif ffi.istype(char_ptr_t, buf) then
+        bufp[0] = buf
+    end
+
+    local c = bufp[0][0]
+    bufp[0] = bufp[0] + 1
+    if (mp_typeof(c) ~= builtin.MP_ARRAY) then
+        error(string.format("%s: unexpected msgpack type '0x%x'",
+                            'decode_array_header_compat', c))
+    end
+
+    local len = decode_array_header_slowly(c, bufp)
+
+    return len, ffi.cast(ffi.typeof(buf), bufp[0])
+end
+
+local function decode_map_header_slowly(c, data)
+    -- fixmap: 0x80 .. 0x8f
+    if bit.band(c, 0xf0) == 0x80 then
+        return bit.band(c, 0xf)
+    end
+
+    -- map 16
+    if c == 0xde then
+        return decode_u16(data)
+    -- map 32
+    elseif c == 0xdf then
+        return decode_u32(data)
```

Comment:
A bit hacky way to avoid reimplementation of the decode_u16 / decode_u32 functions (untested, just for the idea):

```lua
if c == 0xde then -- array 16
    local buf = buffer.ibuf()
    local p = buf:alloc(5)
    p[0] = '\xcd' -- uint 16
    ffi.copy(p + 1, data + 1, 4)
    local res = msgpackffi.decode(p)
    buf:reset()
elseif <...>
```